### ISSUE TITLE
Spack doesn't reuse spec when concretizing

### ISSUE THREAD

**Issue by davide-q** [LLNL/spack#32594]
### Steps to reproduce

```
[I have no name!@co49svnode03 spack_env_clone_nonlock]$ spack spec libbsd
Input spec
--------------------------------
libbsd

Concretized
--------------------------------
libbsd@0.11.5%gcc@12.1.0 arch=linux-rhel8-icelake
    ^libmd@1.0.4%gcc@12.1.0 arch=linux-rhel8-icelake

[I have no name!@co49svnode03 spack_env_clone_nonlock]$ spack env list
==> 3 environments
    JH  p3-10-4_g11-3-0  p3-8-13_g12-1-0
[I have no name!@co49svnode03 spack_env_clone_nonlock]$ spacktivate p3-10-4_g11-3-0
[I have no name!@co49svnode03 spack_env_clone_nonlock]$ spack spec libbsd
Input spec
--------------------------------
libbsd

Concretized
--------------------------------
libbsd@0.11.5%gcc@12.1.0 arch=linux-rhel8-icelake
    ^libmd@1.0.4%gcc@12.1.0 arch=linux-rhel8-icelake

[I have no name!@co49svnode03 spack_env_clone_nonlock]$ ls opt/spack/linux-rhel8-icelake/gcc-11.3.0/libbsd-0.11.5-xl5xnw5gqklushaidesyhyprqkiibgmf/
include/  lib/  share/
[I have no name!@co49svnode03 spack_env_clone_nonlock]$ ls var/spack/environments/p3-10-4_g11-3-0/.spack-env/view/lib/libbsd* -l
lrwxrwxrwx 1 1457002254 1457000513 152 Sep  8 17:02 var/spack/environments/p3-10-4_g11-3-0/.spack-env/view/lib/libbsd.a -> '/home/h505495@honeywell.lab/spack_env_clone_nonlock/opt/spack/linux-rhel8-icelake/gcc-11.3.0/libbsd-0.11.5-xl5xnw5gqklushaidesyhyprqkiibgmf/lib/libbsd.a'
lrwxrwxrwx 1 1457002254 1457000513 157 Sep  8 17:02 var/spack/environments/p3-10-4_g11-3-0/.spack-env/view/lib/libbsd-ctor.a -> '/home/h505495@honeywell.lab/spack_env_clone_nonlock/opt/spack/linux-rhel8-icelake/gcc-11.3.0/libbsd-0.11.5-xl5xnw5gqklushaidesyhyprqkiibgmf/lib/libbsd-ctor.a'
lrwxrwxrwx 1 1457002254 1457000513 153 Sep  8 17:02 var/spack/environments/p3-10-4_g11-3-0/.spack-env/view/lib/libbsd.so -> '/home/h505495@honeywell.lab/spack_env_clone_nonlock/opt/spack/linux-rhel8-icelake/gcc-11.3.0/libbsd-0.11.5-xl5xnw5gqklushaidesyhyprqkiibgmf/lib/libbsd.so'
lrwxrwxrwx 1 1457002254 1457000513 155 Sep  8 17:02 var/spack/environments/p3-10-4_g11-3-0/.spack-env/view/lib/libbsd.so.0 -> '/home/h505495@honeywell.lab/spack_env_clone_nonlock/opt/spack/linux-rhel8-icelake/gcc-11.3.0/libbsd-0.11.5-xl5xnw5gqklushaidesyhyprqkiibgmf/lib/libbsd.so.0'*
lrwxrwxrwx 1 1457002254 1457000513 160 Sep  8 17:02 var/spack/environments/p3-10-4_g11-3-0/.spack-env/view/lib/libbsd.so.0.11.5 -> '/home/h505495@honeywell.lab/spack_env_clone_nonlock/opt/spack/linux-rhel8-icelake/gcc-11.3.0/libbsd-0.11.5-xl5xnw5gqklushaidesyhyprqkiibgmf/lib/libbsd.so.0.11.5'*
[I have no name!@co49svnode03 spack_env_clone_nonlock]$

```

### Error message

No error message, but I would expect that spack spec would return `libbsd@0.11.5%gcc@11.3.0` which spack itself compiled and is available in the loaded environment. Instead it insists on `libbsd@0.11.5%gcc@12.1.0`

### Information on your system

[I have no name!@co49svnode03 spack_env_clone_nonlock]$ spack debug report
* **Spack:** 0.18.1 (13e6f87ef6527954b152eaea303841978e83b992)
* **Python:** 3.6.8
* **Platform:** linux-rhel8-icelake
* **Concretizer:** clingo

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have searched the issues of this repo and believe this is not a duplicate
- [X] I have run the failing commands in debug mode and reported the output

**Comment by davide-q** [comment]
This is very vaguely related to #31345 and #31782 @scheibelp

**Comment by scheibelp** [comment]
A couple questions:

* You mention that `libbsd@0.11.5%gcc@11.3.0` is in fact installed. When you activate `p3-10-4_g11-3-0`, if you run `spack find libbsd` does it give you the expected `libbsd`?
* Either way, I assume this is problematic because if you copied `p3-10-4_g11-3-0` to a new spack instance where `p3-8-13_g12-1-0` was installed but `p3-10-4_g11-3-0` was not, that `spack env activate p3-10-4_g11-3-0; spack install` would not give you the desired result. To control that, you can use `require:` like:

```
packages:
  all:
    require: "%gcc@11.3.0"
```

If you have a mixture of compilers, this will not work; you can use preferences but my guess is that reuse optimization would still override it, so I think that would encourage making changes to tweak the relative priority of reuse vs. preferences (or in the meantime, using `require:` on a per-package basis).

**Comment by davide-q** [comment]
> A couple questions:
> 
> * You mention that `libbsd@0.11.5%gcc@11.3.0` is in fact installed. When you activate `p3-10-4_g11-3-0`, if you run `spack find libbsd` does it give you the expected `libbsd`?

Yes, that works correctly. In fact everything works correctly, but the output as described above. I stumbled into this only because I was trying to write a patch for #31345 and noticed it, otherwise I would not ever noticed.

> * Either way, I assume this is problematic because if you copied `p3-10-4_g11-3-0` to a new spack instance where `p3-8-13_g12-1-0` was installed but `p3-10-4_g11-3-0` was not, that `spack env activate p3-10-4_g11-3-0; spack install` would not give you the desired result. 

Not sure what you mean. I am not copying anything. I am creating these two environments from scratch. And then installing their content by either editing the config (with `spack config edit`) or by copying the yaml and lock file from another machine (for the test we discussed at the last concall). Ok, the latter is a copy, but this problem is also in the former when no copy is involved.

> To control that, you can use `require:` like:
> 
> ```
> packages:
>   all:
>     require: "%gcc@11.3.0"
> ```
> 
> If you have a mixture of compilers, this will not work; 

In this case, I could, but in the general case, unfortunately, I can't. Often, there is something that does not compile correctly with the compiler I'd want to use

> you can use preferences but my guess is that reuse optimization would still override it, so I think that would encourage making changes to tweak the relative priority of reuse vs. preferences (or in the meantime, using `require:` on a per-package basis).

Are you saying that this might not get honored? This is what I have in the config

```
  - matrix:
    - [py-numpy, py-scipy, py-cvxpy, py-numba, py-numexpr, py-natsort, py-pandas,
      py-h5py~mpi^hdf5~mpi, py-tqdm, py-matplotlib, py-joblib, py-cvxopt] 
    - ['%gcc@11.3.0']
    - [^python@3.10.4%gcc@11.3.0]
  - python@3.10.4%gcc@11.3.0
```

**Comment by scheibelp** [comment]
>> Either way, I assume this is problematic because if you copied p3-10-4_g11-3-0 to a new spack instance where p3-8-13_g12-1-0 was installed but p3-10-4_g11-3-0 was not, that spack env activate p3-10-4_g11-3-0; spack install would not give you the desired result.

> Not sure what you mean. I am not copying anything.

Apologies: that was just a hypothetical situation which I assume wouldn't work. I didn't mean to imply that's what you were doing.

>> you can use preferences but my guess is that reuse optimization would still override it, so I think that would encourage making changes to tweak the relative priority of reuse vs. preferences (or in the meantime, using require: on a per-package basis).

> Are you saying that this might not get honored? This is what I have in the config

Sorry no (edit: by which I mean that what you are doing *will* work): I was talking about using preferences in `packages.yaml`; your example explicitly defines the specs in the spec list of `spack.yaml`, and those are guaranteed to hold (furthermore you could extend your solution to support multiple compilers - as long as you explicitly set each one, it is guaranteed to be what Spack chooses).

**Comment by alalazo** [comment]
I see this was reported with v0.18.1 I cannot reproduce on a simpler problem with:

* **Spack:** 0.21.0.dev0 (315873cbd329a05f8cb944bcfb8c9ec2866493ff)
* **Python:** 3.8.10
* **Platform:** linux-ubuntu20.04-icelake
* **Concretizer:** clingo

![Screenshot from 2023-06-08 11-24-17](https://github.com/spack/spack/assets/4199709/521a347b-44a9-4956-bb75-5dfa0c8318a3)

I'll be marking this as unreproducible, but please let us know if this is still an issue and we can work to find a reproducer.
